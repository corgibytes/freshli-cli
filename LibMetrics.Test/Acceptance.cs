using System;
using System.Collections.Generic;
using System.IO;
using ApprovalTests;
using ApprovalTests.Reporters;
using LibMetrics.Languages.Php;
using LibMetrics.Languages.Ruby;
using Xunit;

namespace LibMetrics.Test
{
  [UseReporter(typeof(DiffReporter))]
  public class Acceptance
  {
    [Fact]
    public void RubyGemsWithGitHistory()
    {
      ManifestFinder.Register<RubyBundlerManifestFinder>();
      FileHistoryFinder.Register<GitFileHistoryFinder>();

      var runner = new Runner();

      var rubyFixturePath = Fixtures.Path("ruby", "nokotest");
      var results = runner.Run(rubyFixturePath);

      AssertNokoFixtureRubyGemsResults(results);
    }

    [Fact]
    public void RubyGemsWithHistoryViaGitHub() {
      ManifestFinder.Register<RubyBundlerManifestFinder>();
      FileHistoryFinder.Register<GitFileHistoryFinder>();

      var runner = new Runner();

      var repoUrl =
        "https://github.com/corgibytes/libmetrics-fixture-ruby-nokotest";
      var results = runner.Run(repoUrl);

      AssertNokoFixtureRubyGemsResults(results);
    }

    [Fact]
    public void RubyGemsFeedbinHistoryViaGitHub()
    {
      ManifestFinder.Register<RubyBundlerManifestFinder>();
      FileHistoryFinder.Register<GitFileHistoryFinder>();

      var runner = new Runner();

      var repoUrl = "https://github.com/feedbin/feedbin";
      var results = runner.Run(repoUrl);

      var output = new StringWriter();

      foreach (var resultSet in results) {
        if (resultSet.Date <= new DateTime(2020, 01, 01)) {
          output.WriteLine(
            $"{resultSet.Date.ToString("yyyy/MM/dd")}: " +
            $"{resultSet.LibYear.Total.ToString("F3")}"
          );
        }
      }

      Approvals.Verify(output);
    }

    private static void AssertNokoFixtureRubyGemsResults(IList<MetricsResult> results) {
      // nokotest contains a git repository with the following history:
      // Gemfile v1 created on 2017/01/01 references nokogiri 1.7.0
      // Gemfile v2 created on 2018/01/01 references nokogiri 1.8.1
      // Gemfile v3 created on 2019/01/01 references nokogiri 1.9.1

      var output = new StringWriter();

      foreach (var resultSet in results) {
          if (resultSet.Date <= new DateTime(2020, 01, 01)) {
              output.WriteLine(
                  $"{resultSet.Date.ToString("yyyy/MM/dd")}: " +
                  $"{resultSet.LibYear.Total.ToString("F3")}"
              );
          }
      }

      Approvals.Verify(output);
    }

    [Fact]
    public void ComposerWithoutGitHistory()
    {
      ManifestFinder.Register<PhpComposerManifestFinder>();

      var runner = new Runner();

      var phpFixturePath = Fixtures.Path("php", "large");
      var results = runner.Run(
        phpFixturePath,
        asOf: new DateTime(2020, 01, 01));

      var output = new StringWriter();

      foreach (var resultSet in results)
      {
        output.WriteLine($"{resultSet.Date.ToString("yyyy/MM/dd")}: " +
                         $"{resultSet.LibYear.Total.ToString("F3")}");
      }

      var expected = new StringWriter();
      expected.WriteLine("2020/01/01: 6.126");

      Assert.Equal(expected.ToString(), output.ToString());
    }

    [Fact]
    public void DrupalComposerWithoutGitHistory()
    {
      ManifestFinder.Register<PhpComposerManifestFinder>();

      var runner = new Runner();

      var phpFixturePath = Fixtures.Path("php", "drupal");
      var results = runner.Run(
        phpFixturePath,
        asOf: new DateTime(2020, 01, 01));

      var output = new StringWriter();

      foreach (var resultSet in results)
      {
        output.WriteLine($"{resultSet.Date.ToString("yyyy/MM/dd")}: " +
                         $"{resultSet.LibYear.Total.ToString("F3")}");
      }

      var expected = new StringWriter();
      expected.WriteLine("2020/01/01: 3.214");

      Assert.Equal(expected.ToString(), output.ToString());
    }
  }
}
