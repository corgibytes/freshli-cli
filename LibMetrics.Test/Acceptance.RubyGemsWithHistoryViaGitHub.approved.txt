2017/01/01: 0.000
2017/02/01: 0.022
2017/03/01: 0.022
2017/04/01: 0.227
2017/05/01: 0.227
2017/06/01: 0.364
2017/07/01: 1.852
2017/08/01: 1.852
2017/09/01: 1.852
2017/10/01: 2.416
2017/11/01: 2.416
2017/12/01: 2.416
2018/01/01: 0.000
2018/02/01: 0.362
2018/03/01: 0.362
2018/04/01: 0.362
2018/05/01: 0.362
2018/06/01: 0.362
2018/07/01: 0.740
2018/08/01: 0.789
2018/09/01: 0.789
2018/10/01: 0.789
2018/11/01: 1.044
2018/12/01: 1.044
2019/01/01: 0.000
2019/02/01: 0.071
2019/03/01: 0.071
2019/04/01: 0.266
2019/05/01: 0.342
2019/06/01: 0.342
2019/07/01: 0.342
2019/08/01: 0.342
2019/09/01: 0.647
2019/10/01: 0.647
2019/11/01: 0.868
2019/12/01: 0.868
2020/01/01: 0.962
